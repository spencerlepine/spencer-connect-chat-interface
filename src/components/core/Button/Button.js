// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React, { PureComponent } from 'react';
import PT from 'prop-types';
import styled from 'styled-components';
import { Icon, Loader} from 'connect-core';

const globalStyles = ({ button, globals, fonts }) => (`
  white-space: nowrap;
  color: ${button.color};
  cursor: ${button.cursor};
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: ${fonts.medium};
  display: inline-flex;
  align-items: center;
  justify-content: center;
  max-width: 260px;
`);

const normalStyles = ({ button }) => (`
  padding-top: ${button.normal.padding};
  padding-bottom: ${button.normal.padding};
  font-size: ${button.normal.fontSize};
  box-shadow:  ${button.normal.boxShadow};
`);

const smallStyles = ({ button }) => (`
  padding-top: ${button.small.padding};
  padding-bottom: ${button.small.padding};
  font-size: ${button.small.fontSize};
`);

const defaultStyles = ({ button, globals, spacing }) => (`
  background: ${button.default.bg};
  border-color: ${button.default.borderColor};
  
  &:hover {
    background: ${button.default.bgHover};
  }

  &:focus {
    box-shadow: ${globals.boxShadowFocus};
    outline: none;
  }

  &:active{
    border-color: ${button.default.borderColorActive};
  }

  &:disabled {
    ${globals.disabled}
    pointer-events: none;

    &:hover {
      background: ${button.default.bg};
    }
  }
  > img,.loader{
    margin-right: ${spacing.regular};
    position: relative;
    top: 1px;
  }
`);


const applyButtonStyles = ({ button, globals }, type) => (`
  background: ${button[type].bg};
  border-color: ${button[type].borderColor};

  color: ${button[type].color};

  &:hover {
    background: ${button[type].bgHover};
  }

  &:focus {
    box-shadow: ${globals.boxShadowFocus};
    outline: none;
  }

  &:active{
    border-color: ${button[type].borderColorActive};
  }

  &:disabled {
    ${globals.disabled}
    pointer-events: none;

    &:hover {
      background: ${button[type].bg};
    }
  }
`);

export const StyledButton = styled.button`
  ${({ theme }) => theme.typography.base};
  ${({ theme }) => globalStyles(theme)};
  ${({ theme }) => defaultStyles(theme)};
  ${props => props.type && applyButtonStyles(props.theme, props.type)};
  ${props => props.small ? smallStyles(props.theme) : normalStyles(props.theme)};
`;
StyledButton.displayName = 'StyledButton';

const StyledLink = styled.a`
  ${({ theme }) => theme.typography.base};
  ${props => props.type && applyButtonStyles(props.theme, props.type)};
  ${props => props.small ? smallStyles(props.theme) : normalStyles(props.theme)};
  display: inline-block;
  line-height: normal;
  &:visited,
  &:active,
  &:link {
    text-decoration: none;
    ${({ theme }) => globalStyles(theme)};
  }
`;
StyledLink.displayName = 'StyledLink';

export default class Button extends PureComponent {
  static propTypes = {
    type: PT.oneOf(['default', 'primary', 'secondary','secondary_alt', 'tertiary', 'standard']),
    small: PT.bool,
    disabled: PT.bool,
    href: PT.string,
    iconSize: PT.oneOf(['small', 'medium', "mini"]),
    selected: PT.bool
  }

  static defaultProps = {
    type: 'default',
    small: false,
    disabled: false,
    href: undefined,
    loaderColor: '#fff',
    iconSize: "small",
    selected: false
  }

  render() {
    const {
      href,
    } = this.props;

    if (href) {
      return (
        <StyledLink {...this.props} type={undefined} />
      );
    }
    return (
      <StyledButton {...this.props} data-selected={this.props.selected} disabled={this.props.loading || this.props.disabled}>
        {this.props.loading && <Loader color={this.props.loaderColor}/>}
        {this.props.icon && !this.props.loading && <Icon src={this.props.icon} type={this.props.iconSize}/>}
        {this.props.children}
      </StyledButton>
    );
  }
}