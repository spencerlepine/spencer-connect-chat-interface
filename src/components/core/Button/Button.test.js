// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React from 'react';
import Button from './';

describe('<Button />', () => {
  it('should fire callback when onclick is provided and clicked', () => {
    const onclickEvent = jest.fn();
    const wrapper = shallow(<Button onClick={onclickEvent}>Click Me</Button>);

    expect(onclickEvent).toHaveBeenCalledTimes(0);
    wrapper.simulate('click');
    expect(onclickEvent).toHaveBeenCalledTimes(1);
  });

  describe('styles should match snapshots: ', () => {
    test('default', () => {
      const tree = createTree(<Button>Default</Button>);

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('background', 'linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF)');
      expect(tree).toHaveStyleRule('padding-top', '0.45rem');
      expect(tree).toHaveStyleRule('padding-bottom', '0.45rem');
    });

    test('primary', () => {
      const tree = createTree(<Button type="primary">Primary</Button>);

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('background', 'linear-gradient(#538000,#538000 80%,#649a00)');
    });

    test('secondary', () => {
      const tree = createTree(<Button type="primary">Secondary</Button>);

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('background', 'linear-gradient(#538000,#538000 80%,#649a00)');
    });

    test('tertiary', () => {
      const tree = createTree(<Button type="tertiary">Action</Button>);

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('background', 'linear-gradient(#329AD6,#329AD6 80%,#47a4da)');
    });

    test('small', () => {
      const tree = createTree(<Button small>Small</Button>);

      expect(tree).toMatchSnapshot();
      expect(tree).toHaveStyleRule('padding-top', '4px');
      expect(tree).toHaveStyleRule('padding-bottom', '4px');
    });

    test('disabled', () => {
      const tree = createTree(<Button disabled>Disabled</Button>);

      expect(tree).toMatchSnapshot();
    });
  });
});