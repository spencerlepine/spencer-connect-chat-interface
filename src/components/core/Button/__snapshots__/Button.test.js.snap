// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> styles should match snapshots:  default 1`] = `
.c0 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 > img,
.c0 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

<button
  className="c0"
  data-selected={false}
  disabled={false}
  selected={false}
  type="default"
>
  Default
</button>
`;

exports[`<Button /> styles should match snapshots:  disabled 1`] = `
.c0 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 > img,
.c0 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

<button
  className="c0"
  data-selected={false}
  disabled={true}
  selected={false}
  type="default"
>
  Disabled
</button>
`;

exports[`<Button /> styles should match snapshots:  primary 1`] = `
.c0 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#538000,#538000 80%,#649a00);
  border-color: #538000 #538000 #538000;
  color: #FFFFFF;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 > img,
.c0 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c0:hover {
  background: linear-gradient(#649a00,#649a00);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #538000 #538000 #538000;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#538000,#538000 80%,#649a00);
}

<button
  className="c0"
  data-selected={false}
  disabled={false}
  selected={false}
  type="primary"
>
  Primary
</button>
`;

exports[`<Button /> styles should match snapshots:  secondary 1`] = `
.c0 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#538000,#538000 80%,#649a00);
  border-color: #538000 #538000 #538000;
  color: #FFFFFF;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 > img,
.c0 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c0:hover {
  background: linear-gradient(#649a00,#649a00);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #538000 #538000 #538000;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#538000,#538000 80%,#649a00);
}

<button
  className="c0"
  data-selected={false}
  disabled={false}
  selected={false}
  type="primary"
>
  Secondary
</button>
`;

exports[`<Button /> styles should match snapshots:  small 1`] = `
.c0 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  color: undefined;
  padding-top: 4px;
  padding-bottom: 4px;
  font-size: 0.8125rem;
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 > img,
.c0 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

<button
  className="c0"
  data-selected={false}
  disabled={false}
  selected={false}
  type="default"
>
  Small
</button>
`;

exports[`<Button /> styles should match snapshots:  tertiary 1`] = `
.c0 {
  font-size: 1em;
  line-height: 1.465;
  font-weight: normal;
  font-family: Arial,Helvetica,"sans-serif",Helvetica,sans-serif;
  padding: 0;
  white-space: nowrap;
  color: #111111;
  cursor: pointer;
  text-align: center;
  vertical-align: middle;
  border-width: 0px;
  border-style: solid;
  padding-right: 10px;
  padding-left: 10px;
  font-family: Arial,Helvetica,"sans-serif";
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  max-width: 260px;
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
  border-color: #fff #fff #fff;
  background: linear-gradient(#329AD6,#329AD6 80%,#47a4da);
  border-color: #329AD6 #329AD6 #329AD6;
  color: #FFFFFF;
  padding-top: 0.45rem;
  padding-bottom: 0.45rem;
  font-size: 0.875rem;
  box-shadow: 1px 2px 3px 0px rgba(0,0,0,0.1);
}

.c0:hover {
  background: linear-gradient(#fff,#f9f9f9);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #fff #fff #fff;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#FFFFFF,#FFFFFF 80%,#FFFFFF);
}

.c0 > img,
.c0 .loader {
  margin-right: undefined;
  position: relative;
  top: 1px;
}

.c0:hover {
  background: linear-gradient(#47a4da,#5cafde);
}

.c0:focus {
  box-shadow: 0 0 2px 0 #E77600;
  outline: none;
}

.c0:active {
  border-color: #329AD6 #329AD6 #329AD6;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: 0.7;
  outline: none;
  pointer-events: none;
}

.c0:disabled:hover {
  background: linear-gradient(#329AD6,#329AD6 80%,#47a4da);
}

<button
  className="c0"
  data-selected={false}
  disabled={false}
  selected={false}
  type="tertiary"
>
  Action
</button>
`;
